### PR TITLE
add ability to mute contacts

### LINKED ISSUE #867: Feature: Ability to mute contacts
Use Case
-------------
Sometimes the agents might get spam messages through the widget. In order for it to not affect the productivity of the agent, the system should give the ability to temporarily block the contact. 

Acceptance Criteria
--------------
When an agent blocks a conversation,  the conversation should get marked as resolved.  Any new incoming messages to the conversation wouldn't reopen it until the block period is expired. 

There wouldn't be any visual indicator on the widget side, the contact will still be able to keep sending the messages. 

UI Mock
----------

Screenshot 2020-05-21 at 8 16 21 AM

to be refined by : @nithindavid  @pranavrajs 

Implementation Guidelines 
----------------

- Add a constant for the mute period, let's explore with `6 hours`.
- a new API endpoint in conversations controller for blocking/muting the conversation 
  - when the endpoint is hit the conversation is marked as resolved 
  - the conversation id is added to a Redis set which denoted muted conversation 
- Change the existing `reopen_conversation` method in `message.rb` to add additional logic that ensures muted conversations aren't reopened. 
- Front End code to handle the click on mute contact trigger

### PR DESCRIPTION
**Pull request by kadir96**
## Description

This pr adds ability to mute contacts.

I am not a frontend guy and It is my first time using Vue so my frontend logic could be not idiomatic :)

closes #867
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

There are already tests for backend. To test frontend, I followed steps below;

1. Created new conversation.
2. Clicked `More Details` on the conversation
3. Clicked `Mute Contact`
4. Asserted that status of the conversation changes to `resolved`
5. Asserted that `Mute Contact` disappears
6. Asserted that conversation does not reopen if user sends message

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kadir96** [comment]
Codeclimate suggests refactoring the conversations actions but I guess I can't do it properly as I'm not a frontend guy :)

**Comment by pranavrajs** [comment]
@kadir96 Thanks for the PR. ❤️ 

> Codeclimate suggests refactoring the conversations actions but I guess I can't do it properly as I'm not a frontend guy :)

This is okay, you may ignore this, anyway, that module is already bloated, we were planning to split it to smaller modules.

File: app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue [hunk 1]
```diff
@@ -248,4 +258,10 @@ export default {
     padding: 0.2rem;
   }
 }
+
+.contact--mute {
+  color: #ff382d;
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue:263]: Please use color variables from `dashboard/assets/scss/variables.scss`

File: app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue [hunk 2]
```diff
@@ -90,10 +90,14 @@
         icon="ion-clock"
       />
     </div>
+    <a v-show="!currentChat.muted" class="contact--mute" @click="mute">
+      Mute Contact
```
Comment (pranavrajs) [inline app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue:94]: Please add this in i18n files, add a new key (CONTACT_PANEL.MUTE_CONTACT)

File: app/javascript/dashboard/store/modules/conversations/actions.js [hunk 3]
```diff
@@ -215,6 +215,15 @@ const actions = {
       // Handle error
     }
   },
+
+  muteConversation: async ({ commit }, conversationId) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/conversations/actions.js:219]: Please add a test case for this action.

File: app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue (hunk 1 above)
Comment (kadir96) [reply app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue:263]: There is no variable for red other than `alert-color`. Should I use `alert-color` or add a new variable?

File: app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue (hunk 2 above)
Comment (kadir96) [reply app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue:94]: How should I add this for languages other than English? Would adding `"MUTE_CONTACT": ""` suffice?

Comment (pranavrajs) [reply app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue:94]: You don't need to worry about other languages, just don't add the key in those files. The fallback `en` language will be taken in those cases.

File: app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue (hunk 1 above)
Comment (pranavrajs) [reply app/javascript/dashboard/routes/dashboard/conversation/ContactPanel.vue:263]: Alert-color should be okay, we are working on standardizing the colours, so I would suggest using the variables which are already present.

**Comment by kadir96** [comment]
@pranavrajs I made the changes you requested

**Pull request merged**